### ISSUE TITLE
A fast version for data_generator

### ISSUE THREAD

**Issue by liuwei16** [yhenon/keras-frcnn#150]
It is the for loops of calc_rpn function in the original version of data_generator that makes training slower when default anchors are increasing, thus I have write a fast version based on the AnchorTargetLayer in caffe.
[data_generators_fast.py.tar.gz](https://github.com/yhenon/keras-frcnn/files/1235982/data_generators_fast.py.tar.gz)

**Comment by vuhonganh** [comment]
can you provide how many percentage faster approximately?

**Comment by rodrigo2019** [comment]
from .utils.cython_bbox import bbox_overlaps

ImportError: No module named 'keras_frcnn.utils'

**Comment by yhenon** [comment]
Thanks a lot for contributing this! Is this using the bbox_overlaps from ross girchik's fast-rcnn?

**Comment by liuwei16** [comment]
@yhenon yes it is, and we can precompute the default anchors before training starts, it is beneficial especifially when image size are enlarged

**Comment by rodrigo2019** [comment]
@liuwei16  Could you explain how I can use your function?

**Comment by apacha** [comment]
From what I understand, you have to partially copy code from [the original](https://github.com/rbgirshick/py-faster-rcnn) repository, i.e. `bbox.pyx` to use `data_generators_fast.py`. I've been able to use it in my [fork of the repo](https://github.com/apacha/keras-frcnn/blob/e65b1c50746bc9924ee74b5b55d61a30d3963180/keras_frcnn/data_generators_fast.py#L437). 
I did the following things:

- Copy the above-provided python script to my repo
- Copy the [respective code](https://github.com/apacha/keras-frcnn/tree/e65b1c50746bc9924ee74b5b55d61a30d3963180/keras_frcnn/py_faster_rcnn) from py-faster-rcnn repo for the c-impl. of bbox_overlap
- Fixed Makefile and setup.py
- Run make (e.g. with MingW on Windows) inside of the directory `keras_frcnn/py_faster_rcnn` to compile the `bbox.pyx` 
- Align the signatures and precompute the anchors when calling [`get_anchor_gt`](https://github.com/apacha/keras-frcnn/blob/e65b1c50746bc9924ee74b5b55d61a30d3963180/keras_frcnn/data_generators_fast.py#L435)

**Summary**: The provided script is indeed faster than the original implementation but not dramatically. When training with the Pascal VOC 2007+2012 datasets, it reduced time per epoch from ~650s to ~570s at the expense of temporarily blowing up RAM-usage beyond 13GB (I actually don't know, because my PC only has 16 and it started swapping while pre-computing anchors).

**Comment by rodrigo2019** [comment]
@apacha Thank you for your report, I'm trying to speed up the training, I geting around 1000s per epoch using a GTX 1070, my best try is using a pre trained model, even my dataset is completely different, the results converge more quickly.

**Comment by apacha** [comment]
Although I have been able to run and train this, I am getting different results for `calc_rpn` and `calc_rpn2`, which results these errors https://github.com/yhenon/keras-frcnn/issues/51#issuecomment-331620825 when I attempt to use anchor_box_ratios other than [1,1], [1,2], [2,1]. It seems to be faster and actually the network does learn something, but I am not sure if it really works after all. And I would like to use other anchor box ratios, which this implementation does not allow. 

Unfortunately, this repository desperately lacks validation and testing, to see how good the training goes, as specified in other issues, such as https://github.com/yhenon/keras-frcnn/issues/172 and https://github.com/yhenon/keras-frcnn/issues/165

**Comment by yhenon** [comment]
There is a bug in the current repo where the sizes and scales are not right BTW - the total area should be kept at the 3 ratios but they are currently not

**Comment by Tbarowski** [comment]
Can you explain this a bit more, please.
